### PR TITLE
Implement admin client for balancer

### PR DESCRIPTION
**Pull request by dangleptr**
The PR is based on #516 .

Subtask of #515

### CODE REVIEW

File: src/meta/processors/admin/BalancePlan.cpp [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        partTasks[std::make_pair(task.spaceId_, task.partId_)].emplace_back(index++);
+    }
+    buckets_.resize(std::min(tasks_.size(), (size_t)FLAGS_task_concurrency));
+    for (auto it = partTasks.begin(); it != partTasks.end(); it++) {
+        size_t minNum = tasks_.size();
+        int32_t i = 0, minIndex = 0;
+        for (auto& bucket : buckets_) {
+            if (bucket.size() < minNum) {
+                minNum = bucket.size();
+                minIndex = i;
```
Comment (wadeliuyi) [inline src/meta/processors/admin/BalancePlan.cpp]: if no break after set minIndex=i, the minIndex is the largest one that match condition bucket.size() < minNum

File: src/meta/processors/admin/BalancePlan.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+            };  // onError
+        }  // for (auto j = 0; j < buckets_[i].size(); j++)
+    }  // for (auto i = 0; i < buckets_.size(); i++)
+
+    for (auto& bucket : buckets_) {
+        if (!bucket.empty()) {
+            tasks_[bucket[0]].invoke();
+        }
+    }
+    saveInStore(true);
```
Comment (wadeliuyi) [inline src/meta/processors/admin/BalancePlan.cpp]: saveInStore(true) will be called in onFilish, this is the secend called, why?

File: src/meta/processors/admin/BalancePlan.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+         * */
+        FAILED             = 0x04,
+    };
+
+    BalancePlan(BalanceID id, kvstore::KVStore* kv, AdminClient* client)
+        : id_(id)
+        , kv_(kv)
+        , client_(client) {}
+
+    void addTask(BalanceTask task) {
```
Comment (wadeliuyi) [inline src/meta/processors/admin/BalancePlan.h]: why parameter is not BalanceTask &&task?

File: src/meta/processors/admin/Balancer.cpp [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+            auto spaceId = MetaServiceUtils::spaceId(iter->key());
+            spaces.push_back(spaceId);
+            iter->next();
+        }
+    }
+    plan_ = std::make_unique<BalancePlan>(time::TimeUtils::nowInSeconds(), kv_, client_.get());
+    for (auto spaceId : spaces) {
+        auto tasks = genTasks(spaceId);
+        for (auto& task : tasks) {
+            plan_->addTask(std::move(task));
```
Comment (wadeliuyi) [inline src/meta/processors/admin/Balancer.cpp]: why not define a function “addTasks(vector<task>&&)” in plan?

File: src/meta/processors/admin/Balancer.cpp [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    int32_t minLoad = std::floor(avgLoad);
+    int32_t maxLoad = std::ceil(avgLoad);
+    auto hosts = sortedHostsByParts(newHostParts);
+    CHECK_GT(hosts.size(), 1);
+    auto maxPartsHost = hosts.back();
+    auto minPartsHost = hosts.front();
+    auto lastDelta = maxPartsHost.second - minPartsHost.second + 1;
+    while (maxPartsHost.second > maxLoad
+            || minPartsHost.second < minLoad
+            || maxPartsHost.second - minPartsHost.second < lastDelta) {
```
Comment (wadeliuyi) [inline src/meta/processors/admin/Balancer.cpp]: why is less than lastDelta? If current delta is 0, this function will process the loop, but will do nothing because of “partsFrom.size() <= partsTo.size() + 1”

File: src/meta/processors/admin/BalancePlan.cpp (hunk 2 above)
Comment (dangleptr) [reply src/meta/processors/admin/BalancePlan.cpp]: Because we need to update the balancePlan state

File: src/meta/processors/admin/BalancePlan.h (hunk 3 above)
Comment (dangleptr) [reply src/meta/processors/admin/BalancePlan.h]: Because I want to accept const reference and right reference both.

File: src/meta/processors/admin/Balancer.cpp (hunk 4 above)
Comment (dangleptr) [reply src/meta/processors/admin/Balancer.cpp]: A HA.  There are two reasons:
1.  AddTask is designed firstly.
2.  AddTask is more convenient in UTs

File: src/meta/processors/admin/Balancer.cpp (hunk 5 above)
Comment (dangleptr) [reply src/meta/processors/admin/Balancer.cpp]: Because i want the delta be more small after each iteration.  And if noActions happened, the loop will be broken.

File: src/meta/processors/admin/BalancePlan.cpp (hunk 1 above)
Comment (dangleptr) [reply src/meta/processors/admin/BalancePlan.cpp]: minNum is updated each time.

**Comment by dangleptr** [comment]
Jenkins go

File: src/meta/test/AdminClientTest.cpp [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+        });
+        baton.wait();
+        auto peersRet = client->getPeers(0, 1);
+        CHECK(peersRet.ok());
+        auto hosts = std::move(peersRet).value();
+        ASSERT_EQ(3, hosts.size());
+        ASSERT_EQ(HostAddr(localIp, sc2->port_), hosts[0]);
+        ASSERT_EQ(HostAddr(localIp, sc1->port_), hosts[1]);
+        ASSERT_EQ(HostAddr(1, 1), hosts[2]);
+    }
```
Comment (whitewum) [inline src/meta/test/AdminClientTest.cpp]: let's try more than 3 hosts.

File: src/meta/test/AdminClientTest.cpp [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    {
+        LOG(INFO) << "Test removePart...";
+        folly::Baton<true, std::atomic> baton;
+        client->removePart(0, 0, {localIp, sc->port_}).then([&baton](auto&& st) {
+            CHECK(st.ok());
+            baton.post();
+        });
+        baton.wait();
+    }
```
Comment (whitewum) [inline src/meta/test/AdminClientTest.cpp]: what about add part again?
or let's loop the test a few times

File: src/meta/processors/admin/AdminClient.cpp [hunk 8]
```diff
@@ -5,90 +5,333 @@
...
+        auto resp = std::move(t).value();
+        switch (resp.get_code()) {
+            case storage::cpp2::ErrorCode::SUCCEEDED: {
+                p.setValue(Status::OK());
+                return;
+            }
+            case storage::cpp2::ErrorCode::E_LEADER_CHANGED: {
+                if (retry < retryLimit) {
+                    HostAddr leader(resp.get_leader().get_ip(), resp.get_leader().get_port());
+                    int32_t leaderIndex = 0;
```
[the remark this answers is not in the document]
Comment (dangleptr) [reply src/meta/processors/admin/AdminClient.cpp]: The code here is to find the index of leader in hosts.

File: src/meta/processors/admin/AdminClient.cpp [hunk 9]
```diff
@@ -5,90 +5,333 @@
...
+                    return;
+                }
+                p.setValue(Status::Error("Unknown code %d", static_cast<int32_t>(resp.get_code())));
+                return;
+            }
+        }
+    });
+}
+
+nebula::cpp2::HostAddr AdminClient::to(const HostAddr& addr) {
```
[the remark this answers is not in the document]
Comment (dangleptr) [reply src/meta/processors/admin/AdminClient.cpp]: How about toThriftHost ?

File: src/meta/processors/admin/AdminClient.cpp [hunk 10]
```diff
@@ -5,90 +5,333 @@
...
+    nebula::cpp2::HostAddr thriftAddr;
+    thriftAddr.set_ip(addr.first);
+    thriftAddr.set_port(addr.second);
+    return thriftAddr;
+}
+
+StatusOr<std::vector<HostAddr>> AdminClient::getPeers(GraphSpaceID spaceId, PartitionID partId) {
+    CHECK_NOTNULL(kv_);
+    auto partKey = MetaServiceUtils::partKey(spaceId, partId);
```
[the remark this answers is not in the document]
Comment (dangleptr) [reply src/meta/processors/admin/AdminClient.cpp]: Compiler will optimize it.  The partKey will reuse the returned string's allocation.

File: src/meta/processors/admin/AdminClient.cpp [hunk 11]
```diff
@@ -5,90 +5,333 @@
...
+            std::vector<HostAddr> hosts;
+            hosts.resize(partHosts.size());
+            std::transform(partHosts.begin(), partHosts.end(), hosts.begin(), [](const auto& h) {
+                return HostAddr(h.get_ip(), h.get_port());
+            });
+            return hosts;
+        }
+        case kvstore::ResultCode::ERR_KEY_NOT_FOUND:
+            return Status::Error("Key Not Found");
+        default:
```
[the remark this answers is not in the document]
Comment (dangleptr) [reply src/meta/processors/admin/AdminClient.cpp]: Good point.

File: src/meta/processors/admin/BalanceTask.cpp [hunk 12]
```diff
@@ -38,7 +38,7 @@ void BalanceTask::invoke() {
         case Status::CHANGE_LEADER: {
             LOG(INFO) << taskIdStr_ << "Ask the src to give up the leadership.";
             SAVE_STATE();
-            client_->transLeader(spaceId_, partId_, src_, dst_).thenValue([this](auto&& resp) {
+            client_->transLeader(spaceId_, partId_, src_).thenValue([this](auto&& resp) {
```
[the remark this answers is not in the document]
Comment (dangleptr) [reply src/meta/processors/admin/BalanceTask.cpp]: Good catch.  We should use MS here.

File: src/meta/test/AdminClientTest.cpp (hunk 7 above)
Comment (dangleptr) [reply src/meta/test/AdminClientTest.cpp]: AddPart again will be tested on server side.  The UT is just to test client interface.

File: src/meta/test/BalanceIntegrationTest.cpp [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+#include <folly/synchronization/Baton.h>
+#include "meta/processors/admin/Balancer.h"
+#include "meta/test/TestUtils.h"
+#include "storage/test/TestUtils.h"
+#include "fs/TempDir.h"
+
+namespace nebula {
+namespace meta {
+
+TEST(BalanceIntegrationTest, SimpleTest) {
```
[the remark this answers is not in the document]
Comment (dangleptr) [reply src/meta/test/BalanceIntegrationTest.cpp]: Yes.  The BalanceIntegrationTest will go on when server-side is ready.

File: src/meta/test/AdminClientTest.cpp (hunk 6 above)
Comment (dangleptr) [reply src/meta/test/AdminClientTest.cpp]: Yes. We have more situations to test. Let's open another issue to track it.

**Comment by dangleptr** [comment]
Jenkins go

File: src/meta/processors/admin/AdminClient.cpp [hunk 14]
```diff
@@ -5,90 +5,334 @@
...
+                                           bool asLearner) {
     if (injector_) {
         return injector_->addPart();
     }
-    return Status::OK();
+    storage::cpp2::AddPartReq req;
+    req.set_space_id(spaceId);
+    req.set_part_id(partId);
+    req.set_as_learner(asLearner);
+    return getResponse(host, std::move(req), [] (auto client, auto request) {
```
Comment (wadeliuyi) [inline src/meta/processors/admin/AdminClient.cpp]: I can not understand why need move(req) here, the req’s memory is in caller stack, and the parameter req of getResponse passed by value?

File: src/meta/processors/admin/AdminClient.cpp [hunk 15]
```diff
@@ -5,90 +5,334 @@
...
+    storage::cpp2::CatchUpDataReq req;
+    req.set_space_id(spaceId);
+    req.set_part_id(partId);
+    auto ret = getPeers(spaceId, partId);
+    if (!ret.ok()) {
+        return ret.status();
+    }
+    folly::Promise<Status> pro;
+    auto f = pro.getFuture();
+    getResponse(ret.value(), 0, req, [] (auto client, auto request) {
```
Comment (wadeliuyi) [inline src/meta/processors/admin/AdminClient.cpp]: and here why not move(req)?

File: src/meta/processors/admin/AdminClient.cpp (hunk 14 above)
Comment (dangleptr) [reply src/meta/processors/admin/AdminClient.cpp]: Maybe req will be changed in the future.  So use move here.

File: src/meta/processors/admin/AdminClient.cpp (hunk 15 above)
Comment (dangleptr) [reply src/meta/processors/admin/AdminClient.cpp]: A HA.  Good catch!  We should use move.

File: src/meta/processors/admin/AdminClient.cpp [hunk 16]
```diff
@@ -5,90 +5,334 @@
...
+                if (retry < retryLimit) {
+                    HostAddr leader(resp.get_leader().get_ip(), resp.get_leader().get_port());
+                    int32_t leaderIndex = 0;
+                    for (auto& h : hosts) {
+                        if (h == leader) {
+                            break;
+                        }
+                        leaderIndex++;
+                    }
+                    LOG(INFO) << "Return leder change from " << hosts[index]
```
Comment (whitewum) [inline src/meta/processors/admin/AdminClient.cpp]: leader

File: src/meta/processors/admin/AdminClient.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (whitewum) [reply src/meta/processors/admin/AdminClient.cpp]: sounds better

File: src/meta/test/AdminClientTest.cpp [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+TEST(AdminClientTest, RetryTest) {
```
Comment (whitewum) [inline src/meta/test/AdminClientTest.cpp]: Great!

File: src/meta/test/AdminClientTest.cpp (hunk 7 above)
Comment (whitewum) [reply src/meta/test/AdminClientTest.cpp]: Yep. It's simple for the client,  but a valued corned case.

**Comment by dangleptr** [comment]
> Please consider some corner cases: 1.operations failed at the server side, or 2. client crashed and resume to do some job (what should client do).

Yes,  all interfaces here reentrant functions.  When we implement them on server-side, we should keep the principle
